### PR TITLE
fix Issue 21834 - std.numeric.gcd can't handle negative values

### PR DESCRIPTION
**Pull request by ibuclaw**
This is a pretty lame restriction.  Parameters should always be transformed to an unsigned value.

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

**Comment by ibuclaw** [comment]
Eh... you can't signed and unsigned types as parameters for `gcd`, that seems like another oversight.

**Comment by thewilsonator** [comment]
please merge this when you're satisfied with it.

**Comment by ibuclaw** [comment]
> please merge this when you're satisfied with it.

If you're happy with the new signature...

On reviewing this morning, I'd feel more sure if there was a mixed types test added to verify it is handled correctly.

I'll have a small dig around the git history as well to add a few more comments, as why DigitalMars uses Eulers algorithm is beyond me.

File: std/numeric.d [hunk 1]
```diff
@@ -2934,49 +2934,56 @@ Params:
...
             return a;
 
-        import core.bitop : bsf;
-        import std.algorithm.mutation : swap;
+        version (DigitalMars)
+        {
+            while (b)
+            {
+                immutable t = b;
+                b = a % b;
```
Comment (berni44) [inline std/numeric.d:2964]: `gcd(-15,cast(ubyte) 45)` breaks here with "Error: cannot implicitly convert expression `a % cast(int)b` of type `int` to `ubyte`"

**Comment by ibuclaw** [comment]
@thewilsonator @berni44 - rebased on top of other PR refactoring/fixes and added more tests.

**Review by berni44 — APPROVED** [review]

**Pull request merged**